docs/proposal-platforms.md

```diff
@@ -0,0 +1,259 @@
+  regardless of where they come in the order). We propose a different logic for
+  finding matching platform settings:
+  * The pattern will be matched against the full name using
+    [re.fullmatch](https://docs.python.org/3/library/re.html#re.fullmatch)
+    (rather then `re.match` which is currently used for matching hosts).
+  * A hierarchy of platform match expressions from general to specific can be
+    used because config items are processed in the reverse order specified in the
+    file (i.e. bottom up). This is opposite to the method currently used for
+    matching hosts but is required because any additional platforms defined in
+    user config files get added to the end of the list. Currently it is
+    impossible to add hosts defined as regular expressions in the user config if
+    there is already a more generic regular expression defined in the global
+    config.
+  * There will be no special treatment of sections which define a single
+    platform. The first matching section found will be used.
+  * Note that the first reference to a section in the global or user config
+    determines its position in the list. You can repeat a section later in the
+    config and alter some settings but this doesn't change its position.
+
+* The names of all new and existing settings are subject to change until we
+  release cylc 8.0. Initially names will be chosen which match current usage
+  wherever possible. We intend to review all the config settings once we have
+  the key functionality in place.
+
+* The default platform is `localhost`. At cylc 7, any settings for `localhost`
+  act as defaults for all other hosts but we do **not** propose doing this for
+  platforms. The downside is more duplication if you need to override a
+  particular setting for every platform but we think this is clearer and safer.
+
+* If `hosts` is defined then the platform section heading should specify a single
+  platform name - otherwise you are defining lots of identical platforms which
+  makes no sense (although it is relatively harmless so we probably won't try to
+  prevent this or give any warnings if this happens).
+  * As a future enhancement we could support the use of string substitution using
+    [re.sub](https://docs.python.org/3/library/re.html#re.sub) (this would make
+    the deprecation logic more complicated so we don't propose doing this at
+    first). e.g. something like:
+```ini
+[platforms]
+    [[hpcl[12]-bg]]
+        hosts = ("-bg$","")
+        retrieve job logs = True
+        batch system = background
+```
+
+## Example platform usage
+
+These are examples of how the platform may be assigned in the `[runtime]`
+section for a job. They refer to the example platforms shown above.
+
+`platform = desktop01`
+* The simplest example. The job will be run on the host `desktop01`.
+
+`platform = special`
+* There is no platform named `special` defined so this will fail validation.
```

Comment:
This is an interesting point. It means that suites could fail validation depending on the global configuration.

We should make this optional to allow for CI testing etc.